File: src/proto.zig
const std = @import("std");

pub const User = struct {
    id: usize,
    team: usize,
    name: []const u8,
    color: []const u8,
};

pub const Planet = struct {
    id: usize,
    owner: usize,
    ships: f64,
    x: f64,
    y: f64,
    production: f64,
    radius: f64,
};

pub const Fleet = struct {
    id: usize,
    owner: usize,
    source: usize,
    target: usize,
    ships: f64,
    x: f64,
    y: f64,
    radius: f64,
};

pub const EntityType = enum { user, planet, fleet };
pub const Entity = union(EntityType) { user: User, planet: Planet, fleet: Fleet };

pub const SetMetaFields = enum { you, speed, state };
pub const SetMetaMessage = union(SetMetaFields) {
    you: usize,
    speed: f64,
    state: []const u8,
};

pub const SyncEntity = struct {
    id: usize,
    x: ?f64 = null,
    y: ?f64 = null,
    ships: ?f64 = null,
    radius: ?f64 = null,
    owner: ?usize = null,
    target: ?usize = null,
};

const ServerCommand = enum { set, reset, user, planet, fleet, cancel, sync, destroy, tick, results, pong, print, err };
pub const ServerMessage = union(ServerCommand) {
    set: SetMetaMessage,
    reset: void,
    user: User,
    planet: Planet,
    fleet: Fleet,
    cancel: usize,
    sync: []SyncEntity,
    destroy: usize,
    tick: f64,
    results: [][]const u8,
    pong: []const u8,
    print: []const u8,
    err: []const u8,

    pub fn deinit(self: ServerMessage, gpa: std.mem.Allocator) void {
        switch (self) {
            .sync => {
                gpa.free(self.sync);
            },
            .results => {
                gpa.free(self.results);
            },
            else => {},
        }
    }
};

const ServerCommandMap = std.StaticStringMap(ServerCommand).initComptime(.{
    .{ "/SET", ServerCommand.set },
    .{ "/RESET", ServerCommand.reset },
    .{ "/USER", ServerCommand.user },
    .{ "/PLANET", ServerCommand.planet },
    .{ "/FLEET", ServerCommand.fleet },
    .{ "/CANCEL", ServerCommand.cancel },
    .{ "/DESTROY", ServerCommand.destroy },
    .{ "/TICK", ServerCommand.tick },
    .{ "/RESULTS", ServerCommand.results },
    .{ "/PONG", ServerCommand.pong },
    .{ "/PRINT", ServerCommand.print },
    .{ "/ERROR", ServerCommand.err },
});

const MessageParseError = error{
    EmptyMessage,
    OutOfMemory,
    InvalidCommand,
    InvalidSyncId,
    InvalidSetKey,
};

fn take_float(it: anytype) f64 {
    const next: []const u8 = it.next() orelse return 0.0;
    return std.fmt.parseFloat(f64, next) catch return 0.0;
}

fn take_id(it: anytype) usize {
    const next: []const u8 = it.next() orelse return 0;
    return std.fmt.parseInt(usize, next, 10) catch return 0;
}

fn take_str(it: anytype) []const u8 {
    return it.next() orelse "";
}

fn take_tail(gpa: std.mem.Allocator, it: anytype) MessageParseError![][]const u8 {
    var tail = try std.ArrayList([]const u8).initCapacity(gpa, 6);
    defer tail.deinit(gpa);

    while (it.next()) |chunk| {
        try tail.append(gpa, chunk);
    }

    return try tail.toOwnedSlice(gpa);
}

fn parse_set_command(it: anytype) MessageParseError!SetMetaMessage {
    const key = take_str(it);
    if (std.mem.eql(u8, key, "YOU")) {
        return SetMetaMessage{ .you = take_id(it) };
    } else if (std.mem.eql(u8, key, "SPEED")) {
        return SetMetaMessage{ .speed = take_float(it) };
    } else if (std.mem.eql(u8, key, "STATE")) {
        return SetMetaMessage{ .state = take_str(it) };
    } else {
        return MessageParseError.InvalidSetKey;
    }
}

fn parse_sync_command(gpa: std.mem.Allocator, header: []const u8, it: anytype) MessageParseError![]SyncEntity {
    var syncs = try std.ArrayList(SyncEntity).initCapacity(gpa, 8);
    defer syncs.deinit(gpa);

    while (it.next()) |id| {
        var sync = SyncEntity{
            .id = std.fmt.parseInt(usize, id, 10) catch return MessageParseError.InvalidSyncId,
        };
        for (header) |f| {
            switch (f) {
                'X', 'x' => {
                    sync.x = take_float(it);
                },
                'Y', 'y' => {
                    sync.y = take_float(it);
                },
                'S', 's' => {
                    sync.ships = take_float(it);
                },
                'R', 'r' => {
                    sync.radius = take_float(it);
                },
                'O', 'o' => {
                    sync.owner = take_id(it);
                },
                'T', 't' => {
                    sync.target = take_id(it);
                },
                else => return MessageParseError.InvalidCommand,
            }
        }
        try syncs.append(gpa, sync);
    }

    return try syncs.toOwnedSlice(gpa);
}

pub fn parse_server_message(gpa: std.mem.Allocator, line: []u8) MessageParseError!ServerMessage {
    var chunks = std.mem.splitScalar(u8, line, '\t');
    const head = chunks.next() orelse return MessageParseError.EmptyMessage;

    if (ServerCommandMap.get(head)) |cmd| {
        return switch (cmd) {
            .set => ServerMessage{ .set = try parse_set_command(&chunks) },
            .reset => ServerMessage{ .reset = {} },
            .user => ServerMessage{ .user = User{
                .id = take_id(&chunks),
                .name = take_str(&chunks),
                .color = take_str(&chunks),
                .team = take_id(&chunks),
            } },
            .planet => ServerMessage{ .planet = Planet{
                .id = take_id(&chunks),
                .owner = take_id(&chunks),
                .ships = take_float(&chunks),
                .x = take_float(&chunks),
                .y = take_float(&chunks),
                .production = take_float(&chunks),
                .radius = take_float(&chunks),
            } },
            .fleet => ServerMessage{ .fleet = Fleet{
                .id = take_id(&chunks),
                .owner = take_id(&chunks),
                .ships = take_float(&chunks),
                .x = take_float(&chunks),
                .y = take_float(&chunks),
                .source = take_id(&chunks),
                .target = take_id(&chunks),
                .radius = take_float(&chunks),
            } },
            .cancel => ServerMessage{ .cancel = take_id(&chunks) },
            .destroy => ServerMessage{ .destroy = take_id(&chunks) },
            .tick => ServerMessage{ .tick = take_float(&chunks) },
            .results => ServerMessage{ .results = try take_tail(gpa, &chunks) },
            .pong => ServerMessage{ .pong = take_str(&chunks) },
            .print => ServerMessage{ .print = take_str(&chunks) },
            .err => ServerMessage{ .err = take_str(&chunks) },
            .sync => unreachable, // handled by tail
        };
    } else {
        // TODO handle sync messages
        return ServerMessage{ .sync = try parse_sync_command(gpa, head, &chunks) };
    }
}

pub const ClientCommand = enum { login, send, redir, surrender, tock, ping, message };
pub const ClientMessage = union(ClientCommand) {
    login: struct {
        name: []const u8,
        token: []const u8,
    },
    send: struct {
        proportion: f64,
        source: usize,
        target: usize,
    },
    redir: struct {
        source: usize,
        target: usize,
    },
    surrender: void,
    tock: void,
    ping: []const u8,
    message: []const u8,

    pub fn format(self: ClientMessage, writer: *std.Io.Writer) !void {
        switch (self) {
            .login => try writer.print("/LOGIN\t{s}\t{s}", .{ self.login.name, self.login.token }),
            .send => {
                const pct: u8 = @round(std.math.clamp(self.send.proportion * 100.0, 1.0, 100.0));
                try writer.print("/SEND\t{d}\t{d}\t{d}", .{ pct, self.send.source, self.send.target });
            },
            .redir => try writer.print("/REDIR\t{d}\t{d}", .{ self.redir.source, self.redir.target }),
            .surrender => try writer.print("/SURRENDER", .{}),
            .tock => try writer.print("/TOCK", .{}),
            .ping => try writer.print("/TOCK\t{s}", .{self.ping}),
            .message => try writer.print("{s}", .{self.message}),
        }
    }
};

File: src/proto.rs
use std::{iter, str, string};
use thiserror::Error;

#[derive(Debug, Clone)]
#[allow(unused)]
pub struct User {
    pub id: usize,
    pub team: usize,
    pub name: String,
    pub color: String,
}

#[derive(Debug, Clone)]
#[allow(unused)]
pub struct Planet {
    pub id: usize,
    pub owner: usize,
    pub ships: f64,
    pub x: f64,
    pub y: f64,
    pub production: f64,
    pub radius: f64,
}

#[derive(Debug, Clone)]
#[allow(unused)]
pub struct Fleet {
    pub id: usize,
    pub owner: usize,
    pub source: usize,
    pub target: usize,
    pub ships: f64,
    pub x: f64,
    pub y: f64,
    pub radius: f64,
}

#[derive(Debug, Clone)]
pub struct SyncEntity {
    pub id: usize,
    pub x: Option<f64>,
    pub y: Option<f64>,
    pub ships: Option<f64>,
    pub radius: Option<f64>,
    pub owner: Option<usize>,
    pub target: Option<usize>,
}

#[derive(Debug, Clone)]
#[allow(unused)]
pub enum SetMeta {
    You(usize),
    Speed(f64),
    State(String),
}

#[derive(Debug, Clone)]
#[allow(unused)]
pub enum ServerMessage {
    Set(SetMeta),
    Reset,
    User(User),
    Planet(Planet),
    Fleet(Fleet),
    Cancel(usize),
    Sync(Vec<SyncEntity>),
    Destroy(usize),
    Tick(f64),
    Results(Vec<String>),
    Pong(String),
    Print(String),
    Err(String),
}

#[derive(Debug, Clone)]
#[allow(unused)]
pub enum ClientMessage {
    Login {
        name: String,
        token: String,
    },
    Send {
        proportion: f64,
        source: usize,
        target: usize,
    },
    Redir {
        source: usize,
        target: usize,
    },
    Surrender,
    Tock,
    Ping(String),
    Message(String),
}

impl string::ToString for ClientMessage {
    fn to_string(&self) -> String {
        match self {
            ClientMessage::Login { name, token } => format!("/LOGIN\t{}\t{}", name, token),
            ClientMessage::Send {
                proportion,
                source,
                target,
            } => {
                let pct = (proportion * 100.0).round().clamp(1.0, 100.0) as u8;
                format!("/SEND\t{}\t{}\t{}", pct, source, target)
            }
            ClientMessage::Redir { source, target } => format!("/REDIR\t{}\t{}", source, target),
            ClientMessage::Surrender => "/SURRENDER".to_string(),
            ClientMessage::Tock => "/TOCK".to_string(),
            ClientMessage::Ping(s) => format!("/PING\t{}", s),
            ClientMessage::Message(s) => s.clone(),
        }
    }
}

#[derive(Debug, Error)]
pub enum ParseError {
    #[error("empty message")]
    EmptyMessage,
    #[error("invalid command: {0}")]
    InvalidCommand(String),
    #[error("invalid id for sync: {0}")]
    InvalidSyncId(String),
    #[error("invalid key for /SET: {0}")]
    InvalidSetKey(String),
}

type Chunks<'a> = iter::Peekable<str::SplitTerminator<'a, char>>;

fn take_float(chunks: &mut Chunks) -> f64 {
    chunks.next().and_then(|s| s.parse().ok()).unwrap_or(0.0)
}

fn take_id(chunks: &mut Chunks) -> usize {
    chunks.next().and_then(|s| s.parse().ok()).unwrap_or(0)
}

fn take_str(chunks: &mut Chunks) -> String {
    chunks.next().unwrap_or("").to_string()
}

fn parse_set(chunks: &mut Chunks) -> Result<SetMeta, ParseError> {
    match chunks.next().unwrap_or("") {
        "YOU" => Ok(SetMeta::You(take_id(chunks))),
        "SPEED" => Ok(SetMeta::Speed(take_float(chunks))),
        "STATE" => Ok(SetMeta::State(take_str(chunks))),
        other => Err(ParseError::InvalidSetKey(other.to_string())),
    }
}

fn parse_sync(header: &str, chunks: &mut Chunks) -> Result<Vec<SyncEntity>, ParseError> {
    let mut syncs = Vec::new();

    while let Some(raw_id) = chunks.next() {
        let id = raw_id
            .parse::<usize>()
            .map_err(|_| ParseError::InvalidSyncId(raw_id.to_string()))?;

        let mut sync = SyncEntity {
            id,
            x: None,
            y: None,
            ships: None,
            radius: None,
            owner: None,
            target: None,
        };

        for ch in header.chars() {
            match ch {
                'X' | 'x' => sync.x = Some(take_float(chunks)),
                'Y' | 'y' => sync.y = Some(take_float(chunks)),
                'S' | 's' => sync.ships = Some(take_float(chunks)),
                'R' | 'r' => sync.radius = Some(take_float(chunks)),
                'O' | 'o' => sync.owner = Some(take_id(chunks)),
                'T' | 't' => sync.target = Some(take_id(chunks)),
                _ => return Err(ParseError::InvalidCommand(header.to_string())),
            }
        }

        syncs.push(sync);
    }

    Ok(syncs)
}

pub fn parse_server_message(line: &str) -> Result<ServerMessage, ParseError> {
    let mut chunks = line.split_terminator('\t').peekable();
    let head = chunks.next().ok_or(ParseError::EmptyMessage)?;

    match head {
        "/SET" => Ok(ServerMessage::Set(parse_set(&mut chunks)?)),
        "/RESET" => Ok(ServerMessage::Reset),
        "/USER" => Ok(ServerMessage::User(User {
            id: take_id(&mut chunks),
            name: take_str(&mut chunks),
            color: take_str(&mut chunks),
            team: take_id(&mut chunks),
        })),
        "/PLANET" => Ok(ServerMessage::Planet(Planet {
            id: take_id(&mut chunks),
            owner: take_id(&mut chunks),
            ships: take_float(&mut chunks),
            x: take_float(&mut chunks),
            y: take_float(&mut chunks),
            production: take_float(&mut chunks),
            radius: take_float(&mut chunks),
        })),
        "/FLEET" => Ok(ServerMessage::Fleet(Fleet {
            id: take_id(&mut chunks),
            owner: take_id(&mut chunks),
            ships: take_float(&mut chunks),
            x: take_float(&mut chunks),
            y: take_float(&mut chunks),
            source: take_id(&mut chunks),
            target: take_id(&mut chunks),
            radius: take_float(&mut chunks),
        })),
        "/CANCEL" => Ok(ServerMessage::Cancel(take_id(&mut chunks))),
        "/DESTROY" => Ok(ServerMessage::Destroy(take_id(&mut chunks))),
        "/TICK" => Ok(ServerMessage::Tick(take_float(&mut chunks))),
        "/RESULTS" => Ok(ServerMessage::Results(chunks.map(str::to_string).collect())),
        "/PONG" => Ok(ServerMessage::Pong(take_str(&mut chunks))),
        "/PRINT" => Ok(ServerMessage::Print(take_str(&mut chunks))),
        "/ERROR" => Ok(ServerMessage::Err(take_str(&mut chunks))),
        _ => Ok(ServerMessage::Sync(parse_sync(head, &mut chunks)?)),
    }
}
